rpmio/rpmkeyring.cc

```diff
@@ -363,52 +372,59 @@ pgpDigParams rpmPubkeyPgpDigParams(rpmPubkey key)
     return params;
 }
 
-static rpmPubkey findbySig(rpmKeyring keyring, pgpDigParams sig)
+rpmRC rpmKeyringVerifySig2(rpmKeyring keyring, pgpDigParams sig, DIGEST_CTX ctx, rpmPubkey * keyptr)
 {
+    rpmRC rc = RPMRC_FAIL;
     rpmPubkey key = NULL;
+    std::vector<std::pair<int, std::string>> results = {};
 
-    if (keyring && sig) {
+    if (sig && ctx) {
+	char *lints = NULL;
 	rdlock lock(keyring->mutex);
 	auto keyid = key2str(pgpDigParamsSignID(sig));
-	auto item = keyring->keys.find(keyid);
-	if (item != keyring->keys.end()) {
-	    key = item->second;
-	    pgpDigParams pub = key->pgpkey;
+
+	/* Look for verifying key */
+	auto range = keyring->keys.equal_range(keyid);
+
+	for (auto it = range.first; it != range.second; ++it) {
+	    key = it->second;
+
 	    /* Do the parameters match the signature? */
-	    if ((pgpDigParamsAlgo(sig, PGPVAL_PUBKEYALGO)
-                 != pgpDigParamsAlgo(pub, PGPVAL_PUBKEYALGO)) ||
-                memcmp(pgpDigParamsSignID(sig), pgpDigParamsSignID(pub),
-                       PGP_KEYID_LEN)) {
+	    if (pgpDigParamsAlgo(sig, PGPVAL_PUBKEYALGO)
+		!= pgpDigParamsAlgo(key->pgpkey, PGPVAL_PUBKEYALGO))
+	    {
 		key = NULL;
+		continue;
 	    }
-	}
-    }
-    return key;
-}
 
-rpmRC rpmKeyringVerifySig2(rpmKeyring keyring, pgpDigParams sig, DIGEST_CTX ctx, rpmPubkey * keyptr)
-{
-    rpmRC rc = RPMRC_FAIL;
-
-    if (sig && ctx) {
-	pgpDigParams pgpkey = NULL;
-	rpmPubkey key = findbySig(keyring, sig);
+	    rc = pgpVerifySignature2(key->pgpkey, sig, ctx, &lints);
+	    results.push_back({rc, std::string(lints ?: "")});
+	    free(lints);
 
-	if (key)
-	    pgpkey = key->pgpkey;
+	    if (rc == RPMRC_OK)
+		break;
+	}
 
-	/* We call verify even if key not found for a signature sanity check */
-	char *lints = NULL;
-	rc = pgpVerifySignature2(pgpkey, sig, ctx, &lints);
-	if (lints) {
-	    rpmlog(rc ? RPMLOG_ERR : RPMLOG_WARNING, "%s\n", lints);
+	/* No key, sanity check signature*/
+	if (results.empty()) {
+	    rc = pgpVerifySignature2(NULL, sig, ctx, &lints);
+	    results.push_back({rc, std::string(lints ? lints : "")});
 	    free(lints);
 	}
-	if (keyptr) {
-	    *keyptr = pubkeyPrimarykey(key);
+
+	for (auto const & item : results) {
+	    if (((rc == RPMRC_OK) == (item.first == RPMRC_OK))
```

Comment:
Just a side-remark as I think clearing the results on OK is a better solution here, but such code can usually be made far more readable by giving the complex condition a human understandable name. Eg this might be something like

```
bool relevant_result = ((rc == RPMRC_OK) == (item.first == RPMRC_OK));
if (relevant_result && !item.second.empty)
    rpmlog()...
```